sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala

```diff
@@ -477,6 +482,57 @@ trait ColumnResolutionHelper extends Logging with DataTypeErrorsBase {
         assert(q.children.length == 1)
         q.children.head.output
       },
+
+      resolveOnDatasetId = (datasetid: Long, name: String) => {
```

Comment:
can you check out function `tryResolveDataFrameColumns` in this file and see if it matches your expectations?

Reply:
Can you also check https://github.com/apache/spark/pull/43115 and see if it fixes your problem?

Reply:
thanks @cloud-fan will check.. The thing is that when Column is extracted from a DataFrame , then it comes as resolved ( and in cases of bugs, it is incorrect ). But still will check 43115

Reply:
> can you check out function `tryResolveDataFrameColumns` in this file and see if it matches your expectations?

I will check it out but isn't plan Id tag set only when using  Client ?

Reply:
> Can you also check #43115 and see if it fixes your problem?

@cloud-fan I applied the patch to nearly latest master with the tests for Spark-47320 . The new tests are failing along with existing ones, so I suppose the PR #43115  needs to be enhanced

Reply:
> > can you check out function `tryResolveDataFrameColumns` in this file and see if it matches your expectations?
> 
> I will check it out but isn't plan Id tag set only when using Client ?

The tryResolveDataFrameColumns will not help for 2 reasons.. 
1) Plan ID is added only when code path uses SQLConnect
2) There are cases, where AttributeReference is already added as resolved. That resolution may have been incorrect as it is not based on the deduplication plan.

Reply:
the purpose of https://github.com/apache/spark/pull/43115 is to use plan for both Spark Connect and classic Spark.

Reply:
I see. Let me modify the code to use #43115 and some change of mine to see if tryResolveDataFrameColumns can be used.

Reply:
@cloud-fan @peter-toth  I applied the patch #43115 and verified that in df.select , the projection is added with Alias having UnresolvedAttribute.   There are many test failures in DataFrameSelfJoinSuite with that change. So atleast with the current code base of spark with patch #43115 is not sufficient to fix new bugs which are encountered  nor the existing tests pass. If planId has to be used to fix the issue so that the current code of tryResolveDataFrameColumns can fix the problem, then atleast it will need modification & debugging.

Reply:
> the purpose of #43115 is to use plan id for both Spark Connect and classic Spark. That said, no more already-resolved `AttributeReference`

Further checking the PR #43115  I see that it is substituting  DataSet ID for plan ID for the attributes. Which is I suppose fine, though redundant, because metadata already contains the datasetID . However main issue with the current PR #43115  which makes it untesttable I think is, that for resolution using LogicalPlan ID , the LogicalPlan should contain the Tag PLAN_ID, which is not being set as it is not SQLConnect plan. 
And I am not that well versed to set the Plan ID  for usual Spark LogicalPlans, for testing purposes.
But I will see if I can reconcile my code with the logic used in tryResolveDataFrameColumns

Reply:
I have been trying to see how I can integrate #43115 with this PR. One part is deifinitely easily integratable and simplifies the code is to use is to use the DataSet change of adding UnResolvedAttribute. This will simplify the changes in this PR for select and join, etc.
But I am not  comfortable as of now to have that function create UnresolvedAttribute . Instead prefer UnresolvedAttributeWithTag which wraps the original attribute.
This PRs logic  of resolution involves identifying first Join ( the top level Join) and then traversing the two legs till either a leaf or a binary node is reached. It does not recurse down further. I am not sure we need resolution to happen below the top level join as the user is expected to extract column attributes from only the Dataset involved in top level Join.
Secondly the current code of resolution of DataFrame does not use TagId from the UnresolvedAttribute. It uses the Logical Plan's ID. 
3rd The DataSetId being a Set[Long] instead of single PlanId , is more versatile as it avoids the need of recursion, if at all.

Reply:
@peter-toth @cloud-fan I tried to use the idea of #43115 to return UnresolvedAttributeWithTag ( though implications would be same even if I had returned  UnresolvedAttribute) , but it affects many tests as there are assertions on AttributeReference ( which are now UnresolvedAttr)..
I will try to see if planIDResolution code can be used ..

Reply:
@peter-toth @cloud-fan,
Attached is the patch which describes the attempt to use planID resolution logic. But that does not work because the planID resolution logic goes till the full depth and even attempt to early end the recursion does not work, as the comparison of the return values of the recursion results, does not contain sufficient data.  And further attempt to somehow reuse that code, is proving too cumbersome for me.
[patch.txt](https://github.com/apache/spark/files/14622366/patch.txt)

Though I have removed the previously added new class UnresolvedAttributeWithTag.

Reply:
I have 2 notes to the above:
- @ahshahid, the following worked in Spark 3.5 but failes in 4.0 after https://github.com/apache/spark/pull/41347 for the same reason as described in the old https://github.com/apache/spark/pull/45343:
  ```
  test("SPARK-47217: DeduplicateRelations issue 4") {
    Seq(true, false).foreach(fail =>
      withSQLConf(SQLConf.FAIL_AMBIGUOUS_SELF_JOIN_ENABLED.key -> fail.toString) {
        val df = Seq((1, 2)).toDF("a", "b")
        val df2 = df.select(df("a").as("aa"), df("b").as("bb"))
        val df3 = df.select(df("a"), df("b"))
        val df4 = df2.join(df3, df2("bb") === df("b")).select(df2("aa"), df("a")) // `df("a")` doesn't come the the join's direct children, but from it's descendants 
        checkAnswer(df4, Row(1, 1) :: Nil)
      }
    )
  }
  ```
  In this test `df("a")`'s expression id gets deduplicated (in the right side of the join) and so the original expression id doesn't work in the final select. But I think this test case proves that we need `tryResolveDataFrameColumns()` like deep recursion when we try resolving by plan ids.
- @cloud-fan, I think there is different problem with `tryResolveDataFrameColumns()`.
  I did try to use it for "re-resolving" attribute references that became invalid in a quick test: https://github.com/peter-toth/spark/commit/a873c24372b1d87184149bc5e65c96da1b0db879, but a few test cases failed due to some logicalplans can belong to multiple datasets. E.g. if we have:
  ```
  val df = Seq((1, 2)).toDF("a", "b")
  val df2 = df.toDF()
  ``` 
  then the `df` and `df2` shares the same logicalplan instance and we can't store multiple ids in the current `LogicalPlan.PLAN_ID_TAG`.

Reply:
@peter-toth I agree with your analysis. In the current PR,  the approach I had in mind, was to allow columns from only top Joining dataframes for simplicity purposes . The reason for this thinking was :
1) It allows predictable behaviour and easier for user to comprehend the outcome.
2) while resolving does not need to do deep traversals. In case of repeat dataframes , if we reach the leaves, then dataset IDs most likely will clash, so to resolve ambiguity we would have to resort to something like shortest depth etc ( I still use depth but only 1 level deep).

Reply:
yes. PlanId being a single value is a sort of limitation.  Though I would hold back my thoughts because planId usage code , I have not fully grasped, except that SQLConnect code is very sensitive to it.

Reply:
Maybe we could fix this issue with a small change like this: https://github.com/apache/spark/pull/45552